### DISCUSSION TITLE
Support non top-level struct for?

### DISCUSSION THREAD

**Discussion by k-ye** [taichi-dev/taichi#3868] [Ideas]
Wonder what's your opinions if we support non-top-level struct for?

As a start we can probably support this just for dense fields by applying  this  transformation: https://github.com/taichi-dev/taichi/blob/master/taichi/transforms/demote_dense_struct_fors.cpp

**Comment by yuanming-hu** [comment]
I agree that starting with dense is a good idea :-)

Supporting serialized sparse struct for's is much easier than parallel ones, since you only need a stack to do so (similar to tree traversal). I can't easily come up with a case where you need a sparse struct-for at the non-top level, so maybe we can put sparse struct fors at a lower priority.
